### ISSUE TITLE
Refine doesn't simplify piecewise function

### ISSUE THREAD

**Issue by bderrett** [sympy/sympy#9384]
```
In [8]: from sympy import *
In [9]: from sympy.abc import l
In [10]: refine(Piecewise((1, l<0), (0, True)), Q.positive(l))
Out[10]: Piecewise((1, l < 0), (0, True))
```

The output should be zero, like in Mathematica:

```
In[1]:= Piecewise[{{1, l<0}, {0, True}}]
In[3]:= Refine[%, l > 0]
Out[3]= 0
```

**Comment by meghana1995** [comment]
If I'm not mistaken, the case of refining a piecewise function has not been implemented as of now. 
However, handling this case shouldn't be a difficult task. A function def refine_Piecewise(expr, assumptions) could be added and a corresponding entry added to handlers_dict. Inside this function,  the assumption can be converted to a condition and then the eval() function can be invoked. This is just a suggestion. If it seems fine, I will be happy to proceed and implement it. I would like some opinions on this ( on whether or not this approach is reasonable).

**Comment by NickRSearcy** [comment]
I'd like to add to this. I'm getting a similar issue with `Min` (and other such functions):

```
>>> from sympy import var,Min
>>> f = 2.0*Min(1,var('b'))
>>> f.refine(b<1)
2.0*Min(1, b)
```

I'm expecting `2.0*b` because on the range (-oo, 1), `Min(1,b)=b`. I'm guessing refining of multiple `sympy.functions.elementary.miscellaneous.*` functions has not been implemented. I'll be looking at the Sympy source to see if I can offer anything as far as a pull request but as I've never contributed here, I'm probably not the best person to do so. Anyway, I'd be happy to help anyone interested or to receive guidance from anyone that can help.

**Comment by asmeurer** [comment]
It may not be easy, because currently the assumptions have no knowledge of inequalities at all (only things like Q.positive(1 - b)). Once the assumptions can handle things like this better it will be easy to enable refine for Min and Piecewise.

**Comment by NickRSearcy** [comment]
I couldn't tell if you meant that this should work for `Q.positive(1-b)` but just for the record it doesn't:

```
>>> from sympy import var,Min,Q
>>> f = 2.0*Min(1,var('b'))
>>> f.refine(Q.positive(1-b))
2.0*Min(1, b)
```

**Comment by NickRSearcy** [comment]
Is there work on making assumptions handle these cases better? I can only see the master branch. Anyway, it seems like a feature in at least some other CASs, so I'd like to help implement it anyway I can.

**Comment by asmeurer** [comment]
Any existing work will be on a GitHub pull request (you won't see branches here because people create branches on their forks). 

There's no existing work as far as I know. There has been some discussion about it. I'll see if I can find it.

**Comment by asmeurer** [comment]
I can't find it. You can browse the issues here labeled assumptions. 

The basic question is how to make things like `ask(Q.positive(x - 3), Q.positive(x - 4))` work (i.e., given that `x > 4`, we should know that `x > 3`).

**Comment by meghana1995** [comment]
Also if relational assumptions are implemented, it would make refine for piecewise easier. For example: ask(Q.positive(x), Q.is_true(x > 0)) does not give a meaningful result but it the documentation says that it is work in progress.

I would also like to help handle these assumption if there is not work being done on it currently.

**Comment by czgdp1807** [comment]
Update:
The issue is still unresolved.

**Comment by NotWearingPants** [comment]
Update: The issue is now resolved (idk by which commit/PR)

**Comment by oscarbenjamin** [comment]
This was fixed by 21f987d61704aa80c82d5f89f3fb77991a02f784 from #21309. This issue can be closed if a test is added for `refine`.

**Comment by 1e9abhi1e10** [comment]
> This was fixed by [21f987d](https://github.com/sympy/sympy/commit/21f987d61704aa80c82d5f89f3fb77991a02f784) from #21309. This issue can be closed if a test is added for `refine`.

Added the tests in `tests_refine` #24660
